libraries/lib-wx-wrappers/SpinControl.cpp

```diff
@@ -0,0 +1,309 @@
+{
+   mMaxValue = std::max(value, mMinValue);
+   SetValue(mValue);
+   SetupControls();
+}
+
+double SpinControl::GetMaxValue() const
+{
+   return mMaxValue;
+}
+
+void SpinControl::SetStep(double step)
+{
+   mStep = step;
+   SetupControls();
+}
+
+double SpinControl::GetStep() const
+{
+   return mStep;
+}
+
+void SpinControl::SetFractionalAllowed(bool allow)
+{
+   if (mFrarctionalAllowed == allow)
+      return;
+
+   mFrarctionalAllowed = allow;
+   SetupControls();
+}
+
+bool SpinControl::GetFractionalAllowed()
+{
+   return mFrarctionalAllowed;
+}
+
+void SpinControl::SetName(const TranslatableString& name)
+{
+#if wxUSE_ACCESSIBILITY
+   if (mWindowAccessible == nullptr)
+   {
+      mWindowAccessible = safenew SpinControlAx(this, mTextControl);
+      mTextControl->SetAccessible(mWindowAccessible);
+   }
+
+   mWindowAccessible->SetName(name);
+#endif
+}
+
+void SpinControl::UpdatePrefs()
+{  
+}
+
+void SpinControl::CreateUI()
+{
+   mTextControl = safenew wxTextCtrl(this, wxID_ANY);
+   const auto editorHeight = mTextControl->GetSize().y;
+
+   mSpinButton = safenew wxSpinButton(this);
+   mSpinButton->SetMaxSize({ -1, editorHeight });
+
+   // SpinButton is only used to generate the events,
+   // so keep the value between min (0 by default) and max (100 by default)
+   mSpinButton->SetValue(50);
+
+   auto boxSizer = safenew wxBoxSizer(wxHORIZONTAL);
+
+   boxSizer->Add(mTextControl, wxSizerFlags().Border(wxALL, 0));
+   boxSizer->Add(mSpinButton, wxSizerFlags().Border(wxALL, 0));
+
+   const auto width = GetSize().x;
+
+   if (width > 0)
+   {
+      auto spinWidth = mSpinButton->GetSize().x;
+      const auto editorWidth = std::max(spinWidth, width - spinWidth);
+      mTextControl->SetMaxSize({ editorWidth, editorHeight });
+   }
+
+   SetSizer(boxSizer);
+   Layout();
+
+   Bind(
+      wxEVT_SET_FOCUS,
+      [this](auto& evt)
+      {
+         mTextControl->SetFocus();
+         evt.Skip();
+      });
+
+   Bind(wxEVT_CHAR_HOOK, &SpinControl::OnCharHook, this);
+
+   mTextControl->Bind(
+      wxEVT_KILL_FOCUS,
+      [this](auto& evt)
+      {
+         CommitTextControlValue();
+         evt.Skip();
+      });
+
+   mTextControl->Bind(
+      wxEVT_MOUSEWHEEL,
+      [this](auto& evt)
+      {
+         const auto delta = evt.GetWheelDelta();
+         const auto rotation = evt.GetWheelRotation();
+
+         if (rotation >= delta)
+            DoSteps(evt.ShiftDown() ? 10 : 1);
+         else if (rotation <= delta)
+            DoSteps(evt.ShiftDown() ? -10 : -1);
+      });
+
+   mSpinButton->Bind(
+      wxEVT_SPIN_UP,
+      [this](auto& evt)
+      {
+         DoSteps(1);
+         evt.Veto();
+      });
+
+   mSpinButton->Bind(
+      wxEVT_SPIN_DOWN,
+      [this](auto& evt)
+      {
+         mSpinButton->SetValue(1);
+         DoSteps(-1);
+         evt.Veto();
+      });
+}
+
+void SpinControl::SetupControls()
+{
+   if (mFrarctionalAllowed)
+   {
+      auto validator = wxFloatingPointValidator<ValueType>(
```

Comment:
Have you seen valnum.h?  Our own adapted versions of these classes for reasons I don’t know.

Reply:
Looks to be a copy from the wx source tree. I am confused.

The code comes from 2013 with a message 

``` Backported wxWidgets-3.0rc1 number validators

The are pretty darn slick.  There's an integer one and a floating point 
one.  They support automatic range limiting (ex., you can't even type a 
number outside of the range), proper number formats (ex., you can't 
enter a decimal point in an integer field), you can't enter bogus 
numbers like "0.3-.2", thousands separators are supported, decimal 
precision may be specified and proper number formatting for string 
values (or automatic conversion to int, double, float, etc.).```

I think they match wx validators exactly

Reply:
Was it at that point a future version of wxWidgets?